[Bug]: App reliably crashes on click of 'Default Locale' in the settings menu

### Steps to reproduce

Go to the settings menu, click on 'Default Locale' 

### Expected behaviour

No crash

### Actual behaviour

Crash

### Device name

Xiaomi Redmi Note 10a

### Android version

Android 13

### GreenStash version

3.6.0

### Device logs

_No response_

### Additional information

_No response_

> > WhatsApp.Video.2024-05-11.at.18.02.46.mp4
> > Compression helps too
> 
> Can you click on "view summary" and send me that ss or either copy paste it if it's possible? Because there it'll show actual logs of crash.

java.lang.RuntimeException: Unable to start activity ComponentInfo{com.android.settings/com.android.settings.localepicker.AppLocalePickerActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ActionBar.setDisplayHomeAsUpEnabled(boolean)' on a null object reference
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3834)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3980)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:101)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2394)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:211)
	at android.os.Looper.loop(Looper.java:300)
	at android.app.ActivityThread.main(ActivityThread.java:8296)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:559)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:954)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ActionBar.setDisplayHomeAsUpEnabled(boolean)' on a null object reference
	at com.android.settings.localepicker.AppLocalePickerActivity.onCreate(AppLocalePickerActivity.java:82)
	at android.app.Activity.performCreate(Activity.java:8523)
	at android.app.Activity.performCreate(Activity.java:8487)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1438)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3815)
	... 12 more

> > > WhatsApp.Video.2024-05-11.at.18.02.46.mp4
> > > Compression helps too
> > 
> > 
> > Can you click on "view summary" and send me that ss or either copy paste it if it's possible? Because there it'll show actual logs of crash.
> 
> java.lang.RuntimeException: Unable to start activity ComponentInfo{com.android.settings/com.android.settings.localepicker.AppLocalePickerActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ActionBar.setDisplayHomeAsUpEnabled(boolean)' on a null object reference at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3834) at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3980) at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:101) at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2394) at android.os.Handler.dispatchMessage(Handler.java:106) at android.os.Looper.loopOnce(Looper.java:211) at android.os.Looper.loop(Looper.java:300) at android.app.ActivityThread.main(ActivityThread.java:8296) at java.lang.reflect.Method.invoke(Native Method) at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:559) at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:954) Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ActionBar.setDisplayHomeAsUpEnabled(boolean)' on a null object reference at com.android.settings.localepicker.AppLocalePickerActivity.onCreate(AppLocalePickerActivity.java:82) at android.app.Activity.performCreate(Activity.java:8523) at android.app.Activity.performCreate(Activity.java:8487) at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1438) at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3815) ... 12 more

Damn, MIUI is going to be MIUI, I guess. Even if it's one of the [highlighting features of Android 13](https://developer.android.com/guide/topics/resources/app-languages), Xiaomi is out there, busy breaking standard Android APIs again... Although i use xiaomi myself but my device is currently running hyper os with android 14 where this seems to have fixed.

> > > > WhatsApp.Video.2024-05-11.at.18.02.46.mp4
> > > > Compression helps too
> > > 
> > > 
> > > Can you click on "view summary" and send me that ss or either copy paste it if it's possible? Because there it'll show actual logs of crash.
> > 
> > 
> > java.lang.RuntimeException: Unable to start activity ComponentInfo{com.android.settings/com.android.settings.localepicker.AppLocalePickerActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ActionBar.setDisplayHomeAsUpEnabled(boolean)' on a null object reference at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3834) at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3980) at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:101) at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2394) at android.os.Handler.dispatchMessage(Handler.java:106) at android.os.Looper.loopOnce(Looper.java:211) at android.os.Looper.loop(Looper.java:300) at android.app.ActivityThread.main(ActivityThread.java:8296) at java.lang.reflect.Method.invoke(Native Method) at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:559) at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:954) Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ActionBar.setDisplayHomeAsUpEnabled(boolean)' on a null object reference at com.android.settings.localepicker.AppLocalePickerActivity.onCreate(AppLocalePickerActivity.java:82) at android.app.Activity.performCreate(Activity.java:8523) at android.app.Activity.performCreate(Activity.java:8487) at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1438) at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3815) ... 12 more
> 
> Damn, MIUI is going to be MIUI, I guess. Even if it's one of the [highlighting features of Android 13](https://developer.android.com/guide/topics/resources/app-languages), Xiaomi is out there, busy breaking standard Android APIs again... Although i use xiaomi myself but my device is currently running hyper os with android 14 where this seems to have fixed.

Yeah I know, but I mean what would Android dev be if you didn't need to handle hundreds of edge cases for each device manufacturer 🙃

--- a/app/src/main/java/com/starry/greenstash/ui/screens/settings/composables/SettingsScreen.kt
+++ b/app/src/main/java/com/starry/greenstash/ui/screens/settings/composables/SettingsScreen.kt
@@ -25,6 +25,7 @@
 
 package com.starry.greenstash.ui.screens.settings.composables
 
+import android.annotation.SuppressLint
 import android.content.Intent
 import android.net.Uri
 import android.os.Build
@@ -332,6 +333,7 @@ private fun ThemePickerDialog(
     }
 }
 
+@SuppressLint("InlinedApi")
 @Composable
 private fun LocaleSettings(viewModel: SettingsViewModel) {
     val context = LocalContext.current
@@ -360,8 +362,14 @@ private fun LocaleSettings(viewModel: SettingsViewModel) {
 
     SettingsContainer {
         SettingsCategory(title = stringResource(id = R.string.locales_setting_title))
-        // App locale setting is only available on Android 13+.
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+        // App locale setting is only available on Android 13+
+        // Also, it's not functional on MIUI devices even on Android 13,
+        // Thanks to Xiaomi's broken implementation of standard Android APIs.
+        // See: https://github.com/Pool-Of-Tears/GreenStash/issues/130 for more information.
+        val shouldShowAppLocaleSetting = remember {
+            Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU && !Utils.isMiui()
+        }
+        if (shouldShowAppLocaleSetting) {
             SettingsItem(title = stringResource(id = R.string.app_locale_setting),
                 description = stringResource(id = R.string.app_locale_setting_desc),
                 icon = Icons.Filled.Language,

--- a/app/src/main/java/com/starry/greenstash/utils/Utils.kt
+++ b/app/src/main/java/com/starry/greenstash/utils/Utils.kt
@@ -160,7 +160,7 @@ object Utils {
     }
 
     /**
-     * heck if the device is running on MIUI.
+     * Check if the device is running on MIUI.
      *
      * By default, HyperOS is excluded from the check.
      * If you want to include HyperOS in the check, set excludeHyperOS to false.

--- a/app/src/main/java/com/starry/greenstash/utils/Utils.kt
+++ b/app/src/main/java/com/starry/greenstash/utils/Utils.kt
@@ -33,6 +33,9 @@ import android.os.Build
 import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG
 import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_WEAK
 import androidx.biometric.BiometricManager.Authenticators.DEVICE_CREDENTIAL
+import java.io.BufferedReader
+import java.io.IOException
+import java.io.InputStreamReader
 import java.math.RoundingMode
 import java.text.DecimalFormat
 import java.text.DecimalFormatSymbols
@@ -43,12 +46,14 @@ import java.util.Currency
 import java.util.Locale
 import java.util.TimeZone
 
+
 /**
  * A collection of utility functions.
  */
 object Utils {
 
-    /** Get validated number from the text.
+    /**
+     * Get validated number from the text.
      *
      * @param text The text to validate
      * @return The validated number
@@ -69,7 +74,8 @@ object Utils {
         }
     }
 
-    /** Round the decimal number to two decimal places.
+    /**
+     * Round the decimal number to two decimal places.
      *
      * @param number The number to round
      * @return The rounded number
@@ -81,7 +87,8 @@ object Utils {
         return df.format(number).toDouble()
     }
 
-    /** Format currency based on the currency code.
+    /**
+     * Format currency based on the currency code.
      *
      * @param amount The amount to format
      * @param currencyCode The currency code
@@ -121,7 +128,8 @@ object Utils {
     }
 
 
-    /** Get the epoch time from the LocalDateTime.
+    /**
+     * Get the epoch time from the LocalDateTime.
      *
      * @param dateTime The LocalDateTime object
      * @return The epoch time
@@ -135,7 +143,8 @@ object Utils {
         return dateTime.atZone(timeZone).toInstant().toEpochMilli()
     }
 
-    /** Open the web link in the browser.
+    /**
+     * Open the web link in the browser.
      *
      * @param context The context
      * @param url The URL to open
@@ -149,5 +158,33 @@ object Utils {
             exc.printStackTrace()
         }
     }
+
+    /**
+     * heck if the device is running on MIUI.
+     *
+     * By default, HyperOS is excluded from the check.
+     * If you want to include HyperOS in the check, set excludeHyperOS to false.
+     *
+     * @param excludeHyperOS Whether to exclude HyperOS
+     * @return True if the device is running on MIUI, false otherwise
+     */
+    fun isMiui(excludeHyperOS: Boolean = true): Boolean {
+        val isMiui = !getProperty("ro.miui.ui.version.name").isNullOrBlank()
+        val isHyperOS = !getProperty("ro.mi.os.version.name").isNullOrBlank()
+        return isMiui && (!excludeHyperOS || !isHyperOS)
+    }
+
+    // Private function to get the property value from build.prop.
+    private fun getProperty(property: String): String? {
+        return try {
+            Runtime.getRuntime().exec("getprop $property").inputStream.use { input ->
+                BufferedReader(InputStreamReader(input), 1024).readLine()
+            }
+        } catch (e: IOException) {
+            e.printStackTrace()
+            null
+        }
+    }
+
 
 }

